p2p/protocol/holepunch/holepuncher.go

```diff
@@ -0,0 +1,275 @@
+	refCount sync.WaitGroup
+
+	ids identify.IDService
+
+	// active hole punches for deduplicating
+	activeMx sync.Mutex
+	active   map[peer.ID]struct{}
+
+	closeMx sync.RWMutex
+	closed  bool
+
+	tracer *tracer
+}
+
+func newHolePuncher(h host.Host, ids identify.IDService, tracer *tracer) *holePuncher {
+	hp := &holePuncher{
+		host:   h,
+		ids:    ids,
+		active: make(map[peer.ID]struct{}),
+		tracer: tracer,
+	}
+	hp.ctx, hp.ctxCancel = context.WithCancel(context.Background())
+	h.Network().Notify((*netNotifiee)(hp))
+	return hp
+}
+
+func (hp *holePuncher) beginDirectConnect(p peer.ID) error {
+	hp.closeMx.RLock()
+	defer hp.closeMx.RUnlock()
+	if hp.closed {
+		return ErrClosed
+	}
+
+	hp.activeMx.Lock()
+	defer hp.activeMx.Unlock()
+	if _, ok := hp.active[p]; ok {
+		return ErrHolePunchActive
+	}
+
+	hp.active[p] = struct{}{}
+	return nil
+}
+
+// DirectConnect attempts to make a direct connection with a remote peer.
+// It first attempts a direct dial (if we have a public address of that peer), and then
+// coordinates a hole punch over the given relay connection.
+func (hp *holePuncher) DirectConnect(p peer.ID) error {
+	if err := hp.beginDirectConnect(p); err != nil {
+		return err
+	}
+
+	defer func() {
+		hp.activeMx.Lock()
+		delete(hp.active, p)
+		hp.activeMx.Unlock()
+	}()
+
+	return hp.directConnect(p)
+}
+
+func (hp *holePuncher) directConnect(rp peer.ID) error {
+	// short-circuit check to see if we already have a direct connection
+	for _, c := range hp.host.Network().ConnsToPeer(rp) {
+		if !isRelayAddress(c.RemoteMultiaddr()) {
+			return nil
+		}
+	}
+
+	// short-circuit hole punching if a direct dial works.
+	// attempt a direct connection ONLY if we have a public address for the remote peer
+	for _, a := range hp.host.Peerstore().Addrs(rp) {
+		if manet.IsPublicAddr(a) && !isRelayAddress(a) {
+			forceDirectConnCtx := network.WithForceDirectDial(hp.ctx, "hole-punching")
+			dialCtx, cancel := context.WithTimeout(forceDirectConnCtx, dialTimeout)
+
+			tstart := time.Now()
+			// This dials *all* public addresses from the peerstore.
+			err := hp.host.Connect(dialCtx, peer.AddrInfo{ID: rp})
+			dt := time.Since(tstart)
+			cancel()
+
+			if err != nil {
+				hp.tracer.DirectDialFailed(rp, dt, err)
+				break
+			}
+			hp.tracer.DirectDialSuccessful(rp, dt)
+			log.Debugw("direct connection to peer successful, no need for a hole punch", "peer", rp)
+			return nil
+		}
+	}
+
+	log.Debugw("got inbound proxy conn", "peer", rp)
+
+	// hole punch
+	for i := 0; i < maxRetries; i++ {
+		addrs, rtt, err := hp.initiateHolePunch(rp)
+		if err != nil {
+			log.Debugw("hole punching failed", "peer", rp, "error", err)
+			hp.tracer.ProtocolError(rp, err)
+			return err
+		}
+		synTime := rtt / 2
+		log.Debugf("peer RTT is %s; starting hole punch in %s", rtt, synTime)
+
+		// wait for sync to reach the other peer and then punch a hole for it in our NAT
+		// by attempting a connect to it.
+		timer := time.NewTimer(synTime)
```

Comment:
why are we creating a timer for every iteration? This is no better than using time.After.

Reply:
Because a `time.Timer` can be stopped, whereas a `time.After` cannot.

Reply:
yes but same inefficiency.
We should move the timer out of the loop, defer stop it, and reset for use.

Reply:
Yeah, I don't care about the inefficiency in this case:
1. If the hole punch succeeds, the code only gets executed once.
2. Even if it doesn't, it's 3 timers on the entire connection
Probably not worth getting into trouble with the Reset API.

Reply:
ok fair enough, lets avoid the foot bazooka.